src/axom/slam/ProductSet.hpp

```diff
@@ -34,52 +36,102 @@ namespace slam
  *
  * \see   BivariateSet
  */
-template <typename SetType1 = slam::Set<>, typename SetType2 = slam::Set<>>
+template <typename SetType1 = slam::Set<>,
+          typename SetType2 = slam::Set<>,
+          typename InterfaceType = policies::VirtualInterface>
 class ProductSet final
-  : public BivariateSet<SetType1, SetType2>,
-    RangeSet<typename SetType1::PositionType, typename SetType1::ElementType>
+  : public policies::BivariateSetInterface<InterfaceType,
+                                           SetType1,
+                                           SetType2,
+                                           ProductSet<SetType1, SetType2, InterfaceType>>
 {
+private:
+  using BaseType =
+    policies::BivariateSetInterface<InterfaceType,
+                                    SetType1,
+                                    SetType2,
+                                    ProductSet<SetType1, SetType2, InterfaceType>>;
+
 public:
   using RangeSetType =
     RangeSet<typename SetType1::PositionType, typename SetType1::ElementType>;
   using FirstSetType = SetType1;
   using SecondSetType = SetType2;
-  using BivariateSetType = BivariateSet<FirstSetType, SecondSetType>;
-  using PositionType = typename BivariateSetType::PositionType;
-  using ElementType = typename BivariateSetType::ElementType;
-  using OrderedSetType = typename BivariateSetType::OrderedSetType;
+  using PositionType = typename BaseType::PositionType;
+  using ElementType = typename BaseType::ElementType;
   using ProductSetType = ProductSet<SetType1, SetType2>;
 
-  /** \brief Default constructor */
-  ProductSet() { }
+private:
+  template <typename Dummy, typename SetType>
+  struct RowSet
+  {
+    using Type = SetType;
+    RowSet(PositionType secondSetSize) : m_data(secondSetSize)
+    {
+      //fill in the row data now for getElements(i) function,
+      //since every row is the same, a call to getElements() returns the same set.
+      //
+      // HACK -- this should actually be returning a PositionSet since it always
+      //         goes from 0 to secondSetSize()
+      // This requires a change to the return type of BivariateSet::getElements()
```

Comment:
This workaround/comment was in `ProductSet` before; I think the issue is that we need a common type for the `BivariateSet::getElements()` virtual function declaration that makes sense for both `ProductSet` and `RelationSet`.

In the concrete case, this shouldn't be an issue; we can return a vector-indirection type `OrderedSet` for `RelationSet::getElements()`, and a `RangeSet` for `ProductSet::getElements()`.

But I'm not sure of the best way to deal with this in the virtual case:
- The common base class between the two is just `Set`, so we'd need to return a `const Set&` since `Set` is an abstract base class.
- We'd ideally like to refer to something concrete, so nothing temporarily-constructed. That might be fine for `RelationSet` since we only need to store one "row set", but for `RelationSet`, we need to store each "slice" of the `Relation` array for each index.